skorch/tests/callbacks/test_training.py

```diff
@@ -317,3 +317,276 @@ def test_wrong_threshold_mode(
+
+    def test_initialization_is_effective_pre(self, net_cls, classifier_module,
+                                             classifier_data, initializer):
+        from torch.nn.init import constant_
+        from skorch.utils import to_numpy
+
+        mod = classifier_module()
+
+        net = net_cls(
+            mod,
+            lr=0,
+            max_epochs=1,
+            callbacks=[
+                initializer('dense*.weight', partial(constant_, val=5)),
+                initializer('dense1.bias', partial(constant_, val=10)),
+            ])
+
+        net.fit(*classifier_data)
+
+        assert np.allclose(to_numpy(net.module_.dense0.weight), 5)
+        assert np.allclose(to_numpy(net.module_.dense1.weight), 5)
+        assert np.allclose(to_numpy(net.module_.dense1.bias), 10)
+
+    def test_freezing_is_effective(self, net_cls, classifier_module,
+                                   classifier_data, freezer):
+        from skorch.utils import to_numpy
+
+        net = net_cls(
+            classifier_module,
+            max_epochs=2,
+            callbacks=[
+                freezer('dense*.weight'),
+                freezer('dense1.bias'),
+            ])
+
+        net.initialize()
+
+        assert net.module_.dense0.weight.requires_grad
+        assert net.module_.dense1.weight.requires_grad
+        assert net.module_.dense0.bias.requires_grad
+        assert net.module_.dense1.bias.requires_grad
+
+        dense0_weight_pre = to_numpy(net.module_.dense0.weight).copy()
+        dense1_weight_pre = to_numpy(net.module_.dense1.weight).copy()
+        dense0_bias_pre = to_numpy(net.module_.dense0.bias).copy()
+        dense1_bias_pre = to_numpy(net.module_.dense1.bias).copy()
+
+        # use partial_fit to not re-initialize the module (weights)
+        net.partial_fit(*classifier_data)
+
+        dense0_weight_post = to_numpy(net.module_.dense0.weight).copy()
+        dense1_weight_post = to_numpy(net.module_.dense1.weight).copy()
+        dense0_bias_post = to_numpy(net.module_.dense0.bias).copy()
+        dense1_bias_post = to_numpy(net.module_.dense1.bias).copy()
+
+        assert not net.module_.dense0.weight.requires_grad
+        assert not net.module_.dense1.weight.requires_grad
+        assert net.module_.dense0.bias.requires_grad
+        assert not net.module_.dense1.bias.requires_grad
+
+        assert np.allclose(dense0_weight_pre, dense0_weight_post)
+        assert np.allclose(dense1_weight_pre, dense1_weight_post)
+        assert not np.allclose(dense0_bias_pre, dense0_bias_post)
+        assert np.allclose(dense1_bias_pre, dense1_bias_post)
+
+    def test_freezing_is_effective_pre(self, net_cls, classifier_module,
+                                       classifier_data, freezer):
+        from skorch.utils import to_numpy
+        from torch import tanh
+
+        mod = classifier_module()
+        net = net_cls(
+            mod,
+            max_epochs=2,
+            lr=5,
+            module__nonlin=tanh,
```

Comment:
What is the point of `tanh` and generally, the difference to the previous test?

Reply:
The point is to test also how re-initialization of the module affects the behavior. There is a test that evaluates the functionality. I added a comment explaining just that.